com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java

```diff
@@ -333,6 +336,56 @@ private void mergeSubCollections(IResource owner, CollectionDesc.Builder collect
         collectionBuilder.clearCollectionInstances();
     }
 
+    // Sort instances by hierarchy
+    private List<InstanceDesc.Builder> sortInstancesByHierarchy(IResource resource, List<InstanceDesc.Builder> instances) throws CompileExceptionError {
+        Map<String, InstanceDesc.Builder> instanceMap = new HashMap<>();
+        Map<String, List<String>> dependencies = new HashMap<>();
+
+        for (InstanceDesc.Builder inst : instances) {
+            instanceMap.put(inst.getId(), inst);
+            dependencies.put(inst.getId(), inst.getChildrenList());
+        }
+
+        List<InstanceDesc.Builder> sortedInstances = new ArrayList<>();
+        Set<String> visited = new HashSet<>();
+        Set<String> inStack = new HashSet<>();
+        Stack<String> stack = new Stack<>();
+
+        for (String instanceId : instanceMap.keySet()) {
+            if (!visited.contains(instanceId)) {
+                if (!topologicalSort(instanceId, instanceMap, dependencies, visited, inStack, stack)) {
+                    throw new CompileExceptionError(resource, 0, "Cyclic dependency found in instance hierarchy");
```

Comment:
I decided to add `Cyclic dependency` check here to avoid crashes in the game if developer make a mistake by fixing files manually